### ISSUE TITLE
Specify that writes in `late` local variable initializers disable promotion

### ISSUE THREAD

**Issue by leafpetersen** [dart-lang/language#682]
`late` local variables with initializers are evaluated lazily.  This means that writes in the initializer may be evaluated anywhere the variable is referenced.  We should probably treat this like a write in a closure.  In principle this is more analyzable: we could treat the writes as happening at every reference to the variable, but it's not clear that this is worth the extra mechanism.

cc @johnniwinther @stereotype441

**Comment by lrhn** [comment]
I agree that the initializer of a late local variable should be treated like a closure in every way. It is code that can be executed independently of its location in the source.

The one thing we do know is that no closure can be called before it's declared (being mindful of loops), so we should only treat a variable assigned in a closure/late initializer as unpromotable *after* the closure is created.
(We can also try to figure out what the closure assigns, and if it's compatible with the current promotion, then it's still safe. That would allow 

```dart
Object o = ...; 
if (o is String) {
  list.forEach((v) { 
    o += v; 
  });
}
```
without demoting `o` because of the closure. (It'll not be promotable to anything not a super-type of `String` anywhere after the closure is created, though, because it might become a `String` at any point).

**Comment by leafpetersen** [comment]
@lrhn I don't think that code can be allowed, because we can't guarantee that the closure isn't *called* later, after `o` has been re-assigned to a super type of `String`.

```dart
Object o = ...; 
late Iterable<String> i;
if (o is String) {
  i = list.map((v) { 
    o += v; 
    return o;
  });
}
o = 3;
i.forEach((s) => s.length); // Oops.
```

**Comment by lrhn** [comment]
True, it only works if we know that *from some point onwards* the variable is always at least *T*, and the closure is created after that point and preserves being a *T*. It's non trivial, and perhaps not worth the effort,  but I did run into situations like that while migrating platform libraries.
